dot/rpc/subscription/listeners.go

```diff
@@ -30,121 +30,61 @@ type Listener interface {
 	Listen()
 }
 
-func (c *WSConn) startListener(lid int) {
-	go c.Subscriptions[lid].Listen()
+// WSConnAPI interface defining methors a WSConn should have
+type WSConnAPI interface {
+	safeSend(interface{})
 }
 
-func (c *WSConn) initStorageChangeListener(reqID float64, params interface{}) (int, error) {
-	scl := &StorageChangeListener{
-		Channel: make(chan *state.SubscriptionResult),
-		wsconn:  c,
-	}
-	sub := &state.StorageSubscription{
-		Filter:   make(map[string]bool),
-		Listener: scl.Channel,
-	}
-
-	pA := params.([]interface{})
-	for _, param := range pA {
-		switch p := param.(type) {
-		case []interface{}:
-			for _, pp := range param.([]interface{}) {
-				sub.Filter[pp.(string)] = true
-			}
-		case string:
-			sub.Filter[p] = true
-		default:
-			return 0, fmt.Errorf("unknow parameter type")
-		}
-	}
+// StorageObserver struct to hold data for observer (Observer Design Pattern)
+type StorageObserver struct {
+	id     int
+	filter map[string][]byte
+	wsconn WSConnAPI
+}
 
-	if c.StorageAPI == nil {
-		c.safeSendError(reqID, nil, "error StorageAPI not set")
-		return 0, fmt.Errorf("error StorageAPI not set")
+// Update is called to notify observer of new value
+func (s *StorageObserver) Update(change *state.SubscriptionResult) {
+	if change == nil {
+		return
 	}
 
-	chanID, err := c.StorageAPI.RegisterStorageChangeChannel(*sub)
-	if err != nil {
-		return 0, err
+	result := make(map[string]interface{})
+	result["block"] = change.Hash.String()
+	changes := make([][]string, 0, len(change.Changes))
```

Comment:
Rename `Changes` to `Change` or better, since this represents a single change.

Reply:
updated (using new struct).